File: frontend/js/dedup.js
/**
 * dedup.js
 * Two-layer deduplication for feed items.
 *
 * Layer 1 – within-run: normalise titles and discard near-duplicates that
 *            appear in multiple feeds during the same fetch.
 * Layer 2 – cross-run (session-only): an in-memory Set tracks GUIDs seen
 *            during this page session so auto-poll runs don't re-show cards
 *            already visible on screen. Intentionally NOT persisted to
 *            localStorage — a page reload always shows a fresh feed.
 */

// ── Normalisation ─────────────────────────────────────────────────────────────

/**
 * Produces a collision key from a title for within-run dedup.
 * Strips punctuation/spaces and takes the first 60 chars.
 * @param {string} title
 * @returns {string}
 */
export function normaliseTitle(title) {
  return title
    .toLowerCase()
    .replace(/[^a-z0-9]/g, '')
    .slice(0, 60);
}

// ── Session-scoped seen-GUID helpers ──────────────────────────────────────────

/**
 * Returns a fresh empty Set. Lives only for this page session —
 * reloading the page resets it automatically, so closing and re-opening
 * always shows a full fresh feed.
 * @returns {Set<string>}
 */
export function loadSeenGuids() {
  return new Set();
}

/**
 * No-op: we no longer persist seen GUIDs across sessions.
 * Kept so call-sites in app.js don't need to change.
 * @param {Set<string>} _set
 */
export function saveSeenGuids(_set) {
  // intentionally empty — session-only dedup, nothing to persist
}

/**
 * No-op: nothing to clear since we never write to localStorage.
 */
export function clearSeenGuids() {
  // intentionally empty
}

// ── Main dedup function ───────────────────────────────────────────────────────

/**
 * Deduplicates an array of FeedItems.
 *
 * Returns:
 *  - `fresh`     — items that are genuinely new (pass both layers)
 *  - `dupCount`  — removed in layer 1 (same-run cross-feed duplicates)
 *  - `seenCount` — removed in layer 2 (already rendered this session)
 *
 * Does NOT mutate the seenGuids set — call `markSeen` after classification
 * so we only track items we actually surfaced.
 *
 * @param {import('./fetcher.js').FeedItem[]} items
 * @param {Set<string>} seenGuids
 * @returns {{ fresh: import('./fetcher.js').FeedItem[], dupCount: number, seenCount: number }}
 */
export function deduplicate(items, seenGuids) {
  // Layer 1: within-run dedup by normalised title
  const titleSeen  = new Map();
  const afterLayer1 = [];
  let dupCount = 0;

  for (const item of items) {
    const key = normaliseTitle(item.title);
    if (titleSeen.has(key)) {
      dupCount++;
    } else {
      titleSeen.set(key, true);
      afterLayer1.push(item);
    }
  }

  // Layer 2: session dedup by GUID (in-memory only, resets on page load)
  const fresh = [];
  let seenCount = 0;

  for (const item of afterLayer1) {
    if (seenGuids.has(item.guid)) {
      seenCount++;
    } else {
      fresh.push(item);
    }
  }

  return { fresh, dupCount, seenCount };
}

/**
 * Adds items' GUIDs to the seen set.
 * Call this after classifying and rendering items.
 *
 * @param {import('./fetcher.js').FeedItem[]} items
 * @param {Set<string>} seenGuids  — mutated in place
 */
export function markSeen(items, seenGuids) {
  for (const item of items) seenGuids.add(item.guid);
}

File: src/utils/dedup.py
"""
src/utils/dedup.py
Deduplication helpers for feed items.
Mirrors the logic in frontend/js/dedup.js.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Optional

from src.fetcher.rss import FeedItem

# ── Normalisation ────────────────────────────────────────────────────────────

def normalise_title(title: str) -> str:
    """Collision key: lowercase, alphanumeric only, first 60 chars."""
    return re.sub(r"[^a-z0-9]", "", title.lower())[:60]


# ── File-backed dedup state (CLI equivalent of localStorage) ────────────────

DEFAULT_STATE_FILE = Path(".ai_signal_seen.json")


def load_seen_guids(state_file: Path = DEFAULT_STATE_FILE) -> set[str]:
    """Loads previously seen GUIDs from a JSON file on disk."""
    if not state_file.exists():
        return set()
    try:
        return set(json.loads(state_file.read_text()))
    except Exception:
        return set()


def save_seen_guids(guids: set[str], state_file: Path = DEFAULT_STATE_FILE, cap: int = 500) -> None:
    """Persists the seen-GUID set, capping at `cap` entries."""
    arr = list(guids)[-cap:]
    state_file.write_text(json.dumps(arr, indent=2))


def clear_seen_guids(state_file: Path = DEFAULT_STATE_FILE) -> None:
    """Removes the state file (reset dedup)."""
    state_file.unlink(missing_ok=True)


# ── Main dedup ───────────────────────────────────────────────────────────────

def deduplicate(
    items: list[FeedItem],
    seen_guids: set[str],
) -> tuple[list[FeedItem], int, int]:
    """
    Returns (fresh_items, dup_count, seen_count).

    Layer 1 — within-run: normalised title collision.
    Layer 2 — cross-run: GUID already in seen_guids.
    """
    title_seen: dict[str, bool] = {}
    after_layer1: list[FeedItem] = []
    dup_count = 0

    for item in items:
        key = normalise_title(item.title)
        if key in title_seen:
            dup_count += 1
        else:
            title_seen[key] = True
            after_layer1.append(item)

    fresh: list[FeedItem] = []
    seen_count = 0

    for item in after_layer1:
        if item.guid in seen_guids:
            seen_count += 1
        else:
            fresh.append(item)

    return fresh, dup_count, seen_count


def mark_seen(items: list[FeedItem], seen_guids: set[str]) -> None:
    """Adds item GUIDs to the seen set (mutates in place)."""
    for item in items:
        seen_guids.add(item.guid)
